addons/delivery/tests/test_delivery_cost.py

```diff
@@ -327,3 +327,56 @@ def test_add_carrier_on_picking(self):
             {'product_id': self.product_2.id, 'is_delivery': False, 'product_uom_qty': 1, 'qty_delivered': 1},
             {'product_id': self.normal_delivery.product_id.id, 'is_delivery': True, 'product_uom_qty': 1, 'qty_delivered': 0},
         ])
+
+
+    def test_delivery_cost_gift_card(self):
+        """
+        A customer has a carrier with the amount greater than the one to have
+        free shipping cost, then uses a gift card that lowers that amount to less
+        than the threshold: the shipping cost should still be 0.0
+        """
+
+        if "gift.card" not in self.env:
+            return
+
+        product_delivery_free = self.env['product.product'].create({
+            'name': 'Free Delivery Charges',
+            'type': 'service',
+            'list_price': 40.0,
+            'categ_id': self.env.ref('delivery.product_category_deliveries').id,
+        })
+        free_delivery = self.env['delivery.carrier'].create({
+            'name': 'Delivery Now Free Over 100',
+            'fixed_price': 40,
+            'delivery_type': 'fixed',
+            'product_id': product_delivery_free.id,
+            'free_over': True,
+            'amount': 100,
+        })
+
+
+        sale_normal_delivery_charges = self.SaleOrder.create({
+            'partner_id': self.partner_18.id,
+            'partner_invoice_id': self.partner_18.id,
+            'partner_shipping_id': self.partner_18.id,
+            'pricelist_id': self.pricelist.id,
+            'order_line': [(0, 0, {
+                'name': 'PC Assamble + 2GB RAM',
+                'product_id': self.product_4.id,
+                'product_uom_qty': 1,
+                'product_uom': self.product_uom_unit.id,
+                'price_unit': 120.00,
+            })],
+        })
+        gift_card = self.env['gift.card'].create({
+            'initial_amount': 40,
+        })
+        sale_normal_delivery_charges._pay_with_gift_card(gift_card)
+
+        delivery_wizard = Form(self.env['choose.delivery.carrier'].with_context({
+            'default_order_id': sale_normal_delivery_charges.id,
+            'default_carrier_id': free_delivery.id
+        }))
+        delivery_wizard.save()
```

Comment:
It's missing the confirm, in fact the test passes because the delivery line is not added at all.
```suggestion
        delivery_wizard.save().button_confirm()
```